### PR TITLE
[Design] adding customized model

### PR DESCRIPTION
**Pull request by tonyyang-svail**

### CODE REVIEW

File: doc/customized+model.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+1. TensorFlow 2.x will closely integrate with Keras. [ref](https://www.youtube.com/watch?v=k5c-vg4rjBw)
+
+2. Keras provides more documentation in writing customized models than estimators. For customized estimators, I've only found two examples:
+
+   1. `DNNClassifier` that uses `core_layers.Dense`. [ref](<https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/dnn.py#L200-L226>)
+   2. `DNNClassifier` that uses `tf.layers.dense`. [ref](https://github.com/tensorflow/models/blob/master/samples/core/get_started/custom_estimator.py#L29)
+
+   None of these is suitable for long term development.
+
+3. There are plenty off-the-shelf model repositories written in Keras. To name a few: [ResNet](https://github.com/raghakot/keras-resnet), [Transformer](https://github.com/Lsdefine/attention-is-all-you-need-keras), [Mask  R-CNN](https://github.com/matterport/Mask_RCNN).
```
Comment (wangkuiyi) [inline doc/customized+model.md]: Each of the three URLs doesn't point to a model repository, but a model.

File: Dockerfile [hunk 2]
```diff
@@ -3,7 +3,7 @@ FROM ubuntu:16.04
 RUN apt-get update
 RUN apt-get install -y python3-pip
 RUN pip3 install --upgrade pip
-RUN pip3 install tensorflow mysql-connector-python jupyter sqlflow
+RUN pip3 install tensorflow==2.0.0-alpha0 mysql-connector-python jupyter sqlflow
```
Comment (wangkuiyi) [inline Dockerfile]: What is the reason for switching to 2.0.0-alpha, for that we can use `keras.layers.DenseFeatures`? I noticed that ElasticDL is still using 1.13, which also has Eager Execution. We might need a comment there explaining the reason.

File: doc/customized+model.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+  ```python
+  x = tf.feature_column.input_layer(shape=(5,))
+  for n in hidden_units:
+      x = tf.keras.layers.Dense(n, activation='relu')(x)
+  pred = tf.keras.layers.Dense(n_classes, activation='softmax')(x)
+  model = tf.keras.models.Model(inputs=feature_columns, outputs=pred)
+  ```
+
+  Please be aware that functional API doesn't support feature column as input, not even densed tensors generated from `tf.keras.layers.DenseFeatures(feature_columns)`. [ref1](https://github.com/tensorflow/tensorflow/issues/27416), [ref2](https://stackoverflow.com/questions/54375298/how-to-use-tensorflow-feature-columns-as-input-to-a-keras-model).
```
Comment (wangkuiyi) [inline doc/customized+model.md]: Thank you for this informative point. @yuyicg @skydoorkai @zou000   please have a look at @tonyyang-svail 's point.

**Review by wangkuiyi — DISMISSED** [review]
Thank you, @tonyyang-svail for this great survey and design!

I agree that we go for the keras.Model-derived classes.

**Comment by wangkuiyi** [comment]
I approved this PR, but cannot merge it due to the following CI errors:

```
--- FAIL: TestExecutorTrainAndPredictDNN (2.88s)
    executor_test.go:46: 
        	Error Trace:	executor_test.go:46
        	            				executor_test.go:42
        	Error:      	Should be true
        	Test:       	TestExecutorTrainAndPredictDNN
    executor_test.go:51: 
        	Error Trace:	executor_test.go:51
        	            				executor_test.go:42
        	Error:      	Should be true
        	Test:       	TestExecutorTrainAndPredictDNN
=== RUN   TestExecutorTrainAndPredictBoostedTree
--- FAIL: TestExecutorTrainAndPredictBoostedTree (2.89s)
    executor_test.go:58: 
        	Error Trace:	executor_test.go:58
        	            				executor_test.go:57
        	Error:      	Should be true
        	Test:       	TestExecutorTrainAndPredictBoostedTree
    executor_test.go:59: 
        	Error Trace:	executor_test.go:59
        	            				executor_test.go:57
        	Error:      	Should be true
```

These errors seem independent with this PR. Not sure why they are there.

File: sql/python/models/example.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+model = DNNClassifier(feature_columns=feature_columns, hidden_units=[10, 10], n_classes=3)
+model.compile(optimizer=model.default_optimizer(), loss=model.default_loss())
+
+is_training = False
+if is_training:
+    model.fit(train_ds, validation_data=val_ds, epochs=model.default_training_epochs(), verbose=0)
+    model.save_weights('my_model.h5')
+    print("Done training.")
+else:
+    model.predict(test_ds)
```
Comment (skydoorkai) [inline sql/python/models/example.py:89]: Because DenseFeature is a new feature in tf 2.0alpha0, it is not mature. model.build(input_shape) cannot accept dict of shapes, which is required by DenseFeature.
I believe when 2.0 releases, it should support:
```
    shape = {'c1': (1,1), 'c2': (1,1), 'c3': (1,1), 'c4': (1,1), 'c5': (1,1)}
    model.build(shape)
```

I modified tf code `tensorflow/python/keras/engine/network.py` to add dict support, in `def build(self, input_shape):`
```
    # hack to add dict
    valid_types = (tuple, list, dict, tensor_shape.TensorShape)
    if not isinstance(input_shape, valid_types):
      raise ValueError('Specified input shape is not one of the valid types. '
                       'Please specify a batch input shape of type tuple or '
                       'list of input shapes. User provided '
                       'input type: {}'.format(type(input_shape)))

    if input_shape and not self.inputs:
      # We create placeholders for the `None`s in the shape and build the model
      # in a Graph. Since tf.Variable is compatible with both eager execution
      # and graph building, the variables created after building the model in
      # a Graph are still valid when executing eagerly.
      if context.executing_eagerly():
        graph = func_graph.FuncGraph('build_graph')
      else:
        graph = backend.get_graph()
      with graph.as_default():
        # hack
        if isinstance(input_shape, dict):
            x = {key: base_layer_utils.generate_placeholders_from_shape(input_shape[key])
               for key in input_shape}
        elif isinstance(input_shape, list):
          x = [base_layer_utils.generate_placeholders_from_shape(shape)
               for shape in input_shape]
        else:
          x = base_layer_utils.generate_placeholders_from_shape(input_shape)
```
After that, using dict of shapes as input for model.build works.

Comment (tonyyang-svail) [reply sql/python/models/example.py:89]: @skydoorkai This workaround looks great. Is it possible to contribute this patch to Tensorflow official repo?

File: Dockerfile (hunk 2 above)
Comment (tonyyang-svail) [reply Dockerfile]: Sorry for the confusion. This PR should only contain the design doc.

To explain the switch, TensorFlow 2.0.0-alpha uses `keras.layers.DenseFeatures`, while TensorFlow 1.13 uses `tf.feature_column.input_layer`, to transform feature columns to dense tensors. As mentioned [here](https://stackoverflow.com/a/54718798), Keras is becoming the de-facto high-level API for TensorFlow. So I would suggest using `keras.layers.DenseFeatures`, which is in TensorFlow 2.0.0-alpha.

Upgrading the Tensorflow version will cause failure in CI. I will remove this line change and file it in a separate PR.

File: doc/customized+model.md (hunk 1 above)
Comment (tonyyang-svail) [reply doc/customized+model.md]: Sure, I have changed "model repository" to "model".

**Comment by tonyyang-svail** [comment]
A survey on weight initialization in Keras: https://github.com/sql-machine-learning/sqlflow/wiki/Study-Keras:-weight-initialization

File: sql/python/models/example.py (hunk 4 above)
Comment (skydoorkai) [reply sql/python/models/example.py:89]: We can file an issue for this. Not sure if the workaround is the right solution for it.

File: doc/customized+model.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+We choose Keras over Estimator for the following reasons:
+
+1. TensorFlow 2.x will closely integrate with Keras. [ref](https://www.youtube.com/watch?v=k5c-vg4rjBw)
+
+2. Keras provides more documentation in writing customized models than estimators. For customized estimators, I've only found two examples:
+
+   1. `DNNClassifier` that uses `core_layers.Dense`. [ref](<https://github.com/tensorflow/estimator/blob/master/tensorflow_estimator/python/estimator/canned/dnn.py#L200-L226>)
+   2. `DNNClassifier` that uses `tf.layers.dense`. [ref](https://github.com/tensorflow/models/blob/master/samples/core/get_started/custom_estimator.py#L29)
+
+   None of these is suitable for long term development.
```
Comment (wangkuiyi) [inline doc/customized+model.md]: Why "not suitable for long term development"? Could you please elaborate?

File: doc/customized+model.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      def call(self, inputs):
+          x = self.feature_layer(inputs)
+          for hidden_layer in self.hidden_layers:
+              x = hidden_layer(x)
+          return self.prediction_layer(x)
+  
+  model = DNNClassifier(feature_columns, hidden_units, n_classes)
+  ```
+
+  Please be aware that models subclassing from `tf.keras.Model` only supports `save_weights` and `load_weights`. [ref1](https://stackoverflow.com/questions/51806852/cant-save-custom-subclassed-model), [ref2](https://stackoverflow.com/questions/52826134/keras-model-subclassing-examples).
```
Comment (wangkuiyi) [inline doc/customized+model.md]: I was confused by the point here. Obviously, subclasses support many more than `save_weights` and `load_weights` -- we can see `call` and `__init__`.

After reading the links you provided, I got that you didn't complete your sentence, which seems should have been:

> Please be aware that Keras Model subclasses allow us to save and load model parameters but not model topologies.

File: doc/customized+model.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  ```python
+  class DNNClassifier(tf.keras.Model):
+      def __init__(self, feature_columns, hidden_units, n_classes):
+          super(DNNClassifier, self).__init__()
+          self.feature_layer = tf.keras.layers.DenseFeatures(feature_columns)
+          self.hidden_layers = []
+          for hidden_unit in hidden_units:
+              self.hidden_layers.append(tf.keras.layers.Dense(hidden_unit))
+          self.prediction_layer = tf.keras.layers.Dense(n_classes, activation='softmax')
+
```
Comment (wangkuiyi) [inline doc/customized+model.md:31]: Do we need to add `build` to this example to make it complete for the discussion here?

File: doc/customized+model.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  ```python
+  x = tf.feature_column.input_layer(shape=(5,))
+  for n in hidden_units:
+      x = tf.keras.layers.Dense(n, activation='relu')(x)
+  pred = tf.keras.layers.Dense(n_classes, activation='softmax')(x)
+  model = tf.keras.models.Model(inputs=feature_columns, outputs=pred)
+  ```
+
+  Please be aware that functional API doesn't support feature column as input, not even densed tensors generated from `tf.keras.layers.DenseFeatures(feature_columns)`. [ref1](https://github.com/tensorflow/tensorflow/issues/27416), [ref2](https://stackoverflow.com/questions/54375298/how-to-use-tensorflow-feature-columns-as-input-to-a-keras-model).
```
Comment (wangkuiyi) [inline doc/customized+model.md]: Does this point mean that we cannot call do

```python
x = tf.keras.layers.DenseFeatues(feature_columns)
y = tf.keras.layers.Dense(n, activation='relu')(x)
```

because x is returned by `tf.keras.layers.DenseFeature`?

**Review by wangkuiyi — COMMENTED** [review]
`    `

File: sql/python/models/example.py (hunk 4 above)
Comment (wangkuiyi) [reply sql/python/models/example.py:89]: @skydoorkai could you share your modification? It is good to have it in your personal repo or in an issue in the TensorFlow repo. Anyway, more engineers in the community could understand the requirement and your solution.

**Review by wangkuiyi — APPROVED** [review]

File: sql/python/models/example.py (hunk 4 above)
Comment (skydoorkai) [reply sql/python/models/example.py:89]: I will create a dockerfile in my repo and file an issue in tensorflow repo.

Comment (skydoorkai) [reply sql/python/models/example.py:89]: Put the repro example and the solution here:
https://github.com/skydoorkai/tests_and_issues/tree/master/tensorflow/densefeature

Also filed a tf bug:
https://github.com/tensorflow/tensorflow/issues/28111

Comment (tanzhenyu) [reply sql/python/models/example.py:89]: This workaround looks good to me. Do you want to make this a PR to TF? Or do you want us to add it?
